synapse/tests/test_cortex.py

```diff
@@ -988,6 +988,84 @@ def runtufobydefault(self, core):
+            #(True, ('othername', {'something': 'c'})),
+            #(True, ('splice', {'act': 'z*'})),
+            #(False, ('splice', {'act': 'y*'}))
+        )
+        msgs = (
+            ('splice', {'mesg': ('node:add', {})}),  # allowed because of act
+            #('splice', {'act': 'b', 'form': 'nope'}),  # disallowed because of form
+            #('othername', {'something': 'c', 'form': 'nope'}),  # allowed because of act
+            #('splice', {'act': 'b', 'form': 'yeap'}),  # allowed because of act
+            #('splice', {'act': 'b'}),  # allowed because of act
+            #('splice', {'act': 'something:else'}),  # not in rules, dropped
+            #('splice', {'act': 'zoop'}),  # allowed because act starts with a z
+            #('splice', {'act': 'yes'}),  # disallowed because act starts with a y
+        )
+        expected = [
+            (0, 11, ('node:add', {})),
+            #(22, 56, ('othername', {'something': 'c', 'form': 'nope'})),
+            #(56, 81, ('splice', {'act': 'b', 'form': 'yeap'})),
+            #(81, 96, ('splice', {'act': 'b'})),
+            #(96, 114, ('splice', {'act': 'zoop'}))
+        ]
+
+        # Add a membrane via Cortex API
+        with self.getTestDir() as dirn:
+            with s_cortex.openurl('dir:///' + dirn) as core:
+
+                # spin up a membrane and fire all the messages at the core
+                core.addCoreMembrane(name, rules)
+                [core.dist(msg) for msg in msgs]
+
+                # subscribe to the membrane fifo and check the messages that made it through
+                actual = []
+                core.subCoreFifo(name, actual.append)
+                self.eq(actual, expected)
+
+        # Add membranes via setting config opt
+        with self.getTestDir() as dirn:
+            with s_cortex.openurl('dir:///' + dirn) as core:
+
+                # spin up a membrane and fire all the messages at the core
+                core.setConfOpt('membranes', ((name, rules),))
+                [core.dist(msg) for msg in msgs]
+
+                # subscribe to the membrane fifo and check the messages that made it through
+                actual = []
+                core.subCoreFifo(name, actual.append)
+                self.eq(actual, expected)
+
+        # Test missing and dup membranes
+        with self.getTestDir() as dirn:
+            with s_cortex.openurl('dir:///' + dirn) as core:
+
+                self.raises(NoSuchMembrane, core.delCoreMembrane, name)
+                core.addCoreMembrane(name, rules)
+                self.raises(MembraneExists, core.addCoreMembrane, name, rules)
+
+                [core.dist(msg) for msg in msgs]
+
+                # subscribe to the membrane fifo and check the messages that made it through
+                actual = []
+                core.subCoreFifo(name, actual.append)
+                self.eq(actual, expected)
+
+                # remove and re-add the membrane, subscribe again and make sure it is empty
```

Comment:
@invisig0th @vEpiphyte - this test fails. I spin up a core, add a membrane, fire some messages into it, some of which make it through the membrane and into the fifo. Then I do some assertions on the contents of the fifo. Next I remove the membrane (the entire fifo is deleted in this step) then re-add the membrane and check that its fifo is empty, but somehow the fifo still contains the old messages. I'll do some more debugging.

Reply:
fixed